### PR TITLE
DDF-3494 Pull logic for inserting AudienceRestriction data to SAML assertions to its own plugin

### PR DESCRIPTION
**Pull request by bakejeyner**
DDF-3494 Pull logic for inserting AudienceRestriction data to SAML assertions to its own plugin

#### What does this PR do?
Pull logic for inserting AudienceRestriction data to SAML assertions to its own plugin and adds respective unit tests

#### Who is reviewing it? 
@blen-desta @Lambeaux 

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@clockard
@coyotesqrl 
#### How should this be tested? (List steps with links to updated documentation)
Should have normal login behavior

#### What are the relevant tickets?
[3494](https://codice.atlassian.net/browse/3494)
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  <parent>
+    <groupId>ddf.security.idp</groupId>
+    <artifactId>security-idp-plugin</artifactId>
+    <version>2.12.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>security-idp-plugin-audiencerestriction</artifactId>
+  <name>DDF :: SECURITY :: IDP :: PLUGIN ::</name>
+  <version>2.12.0-SNAPSHOT</version>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Remove; it's inherited and redundant.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>ddf.security.idp</groupId>
+    <artifactId>security-idp-plugin</artifactId>
+    <version>2.12.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>security-idp-plugin-audiencerestriction</artifactId>
+  <name>DDF :: SECURITY :: IDP :: PLUGIN ::</name>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Name does not match our style. Should look like its sibling, so:
`DDF :: Security :: IdP :: Plugin :: Audience Restriction`

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  <version>2.12.0-SNAPSHOT</version>
+
+  <dependencies>
+    <dependency>
+      <groupId>ddf.security.idp</groupId>
+      <artifactId>security-idp-plugin-api</artifactId>
+      <version>2.12.0-SNAPSHOT</version>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Use variable, `${project.version}`

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    <!--Spock dependencies-->
+    <dependency>
+      <groupId>ddf.lib</groupId>
+      <artifactId>spock-shaded</artifactId>
+      <version>${project.version}</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+</project>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Missing `<build>` element with `maven-bundle-plugin` and `jacoco-maven-plugin` plugin configs.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.Audience;
+import org.opensaml.saml.saml2.core.AudienceRestriction;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.impl.AudienceBuilder;
+import org.opensaml.saml.saml2.core.impl.AudienceRestrictionBuilder;
+
+public class AudienceRestrictionPresignPlugin implements SamlPresignPlugin {
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java]: ❗️ Javadoc. Recommend looking at `SubjectConfirmationPresignPlugin` for a guide on the level/type of information to include here.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+import ddf.security.samlp.SamlProtocol
+import org.apache.wss4j.common.saml.OpenSAMLUtil
+import org.opensaml.saml.saml2.core.*
+import spock.lang.Specification
+
+class AudienceRestrictionPresignPluginSpec extends Specification {
+    def ISSUER_ID = "https://localhost:8993/services/idp/login"
+
+    // general mocks
+    AudienceRestrictionPresignPlugin plugin = new AudienceRestrictionPresignPlugin()
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️ I would separate out these first few non-mock objects from the mocks

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    List<String> spMetadata = new ArrayList<>()
+    Set<SamlProtocol.Binding> bindings = new HashSet<>()
+    AuthnRequest authNReq = Mock(AuthnRequest)
+    Issuer issuer = Mock(Issuer)
+
+    // response mocks
+    Response response = Mock(Response)
+    List<Assertion> assertions = new ArrayList<>()
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️ Same: non-mock supporting objects would probably be clearer moved to a separate section from the mock generations.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml (hunk 1 above)
Comment (clockard) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: Can leave this version number out an just inherit the parent version.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml (hunk 2 above)
Comment (clockard) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: Add `Audience Restriction` to the end of the name?

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml (hunk 3 above)
Comment (clockard) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: Use ${project.version} here

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java (hunk 5 above)
Comment (clockard) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java]: Maybe add a little javadoc to this class to tell what it does

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        setup:
+        // add audienceRestriction to audienceRestrictions
+        audienceRestrictions.add(audienceRestriction)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        response.assertions[0].conditions.audienceRestrictions[0].audiences.size() == 3
+        response.assertions[0].conditions.audienceRestrictions[0].audiences[2].audienceURI == ISSUER_ID
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️ This feels a little fragile, depending on the second entry in the audRest. list to be the value you expect. It's a bit cleaner to write this test like this:
```groovy
    def 'add audience to audience restriction with existing audiences'() {
        setup:
        // add audienceRestriction to audienceRestrictions
        audienceRestrictions.add(this.audienceRestriction)

        when:
        plugin.processPresign(response, authNReq, spMetadata, bindings)
        def audienceRestriction = response.assertions[0].conditions.audienceRestrictions[0]

        then:
        audienceRestriction.audiences.size() == 3
        audienceRestriction.audiences*.audienceURI.count {it == ISSUER_ID} == 1
    }
```
(Also note the `audienceRestriction` variable to simplify.)

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    def 'add audience to multiple audience restrictions'() {
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️ Same comments about fragility of looking at specific index elements.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/resources/ExistingAudienceRestrictionWithAudiencesResponse.xml [hunk 10]
```diff
@@ -0,0 +1,120 @@
+<!--
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/resources/ExistingAudienceRestrictionWithAudiencesResponse.xml]: ❓ You're not using this file anywhere?

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy (hunk 7 above)
Comment (bakejeyner) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: I would like to keep it this way; these are in hierarchical order. The top most one requires the second top most one and so on. The only exception is near the bottom where the numbers represent siblings.

Comment (coyotesqrl) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: Understood. Then I suggest at least changing your in-line comments to reflect that this is a heterogenous set of mocks and concrete objects.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml [hunk 11]
```diff
@@ -13,102 +13,102 @@
...
+                  <limits>
+                    <limit>
+                      <counter>INSTRUCTION</counter>
+                      <value>COVEREDRATIO</value>
+                      <minimum>0.93</minimum>
+                    </limit>
+                    <limit>
+                      <counter>BRANCH</counter>
+                      <value>COVEREDRATIO</value>
+                      <minimum>0.59</minimum>
```
Comment (blen-desta) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml]: ❓ Was there a change to the coverage? Why did we put `0.59`?

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        audiences.add(audience2)
+        audiences2.add(audience)
+
+        // stub the audiences to the audienceRestriction
+        audienceRestriction.audiences >> audiences
+        audienceRestriction2.audiences >> audiences2
+        emptyAudienceRestriction.audiences >> emptyAudiences
+    }
+
+    def 'add audience to non existent audience restriction'() {
```
Comment (blen-desta) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ❓ To an empty audience restriction or a non-existing?

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>ddf.security.idp</groupId>
+    <artifactId>security-idp-plugin</artifactId>
+    <version>2.12.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>security-idp-plugin-audiencerestriction</artifactId>
+  <name>DDF :: SECURITY :: IDP :: PLUGIN :: AUDIENCE RESTRICTION</name>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Fix the capitalization to match our standard.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        response.assertions[0].conditions.audienceRestrictions[0].audiences[0].audienceURI == ISSUER_ID
+    }
+
+    def 'add audience to audience restriction with existing audiences'() {
+        setup:
+        audienceRestrictions.add(this.audienceRestriction)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+        def audienceRestriction = response.assertions[0].conditions.audienceRestrictions[0]
```
Comment (blen-desta) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️  You can do the same in the two test above. Put the `audienceRestriction` in a variable and use it in the `then` block. It seems like you do it everywhere else except those two.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/pom.xml]: ❗️ Missing `<packaging>` element:
`<packaging>bundle</packaging>`

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    AudienceRestrictionPresignPlugin plugin = new AudienceRestrictionPresignPlugin()
+
+    // general mocks
+    AuthnRequest authNReq = Mock(AuthnRequest)
+    Issuer issuer = Mock(Issuer)
+
+    List<String> spMetadata = new ArrayList<>()
+    Set<SamlProtocol.Binding> bindings = new HashSet<>()
+
+    // heterogenous set of mocks and concrete objects
```
Comment (blen-desta) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: ✏️  `heterogeneous`?

File: platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml [hunk 17]
```diff
@@ -13,102 +13,102 @@
...
+              <goal>check</goal>
+            </goals>
+            <configuration>
+              <haltOnFailure>true</haltOnFailure>
+              <rules>
+                <rule>
+                  <element>BUNDLE</element>
+                  <limits>
+                    <limit>
+                      <counter>INSTRUCTION</counter>
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml]: ❓ Did you check to see if coverage numbers went up with the transition of the audience restriction work to its own bundle? If so, these can be adjusted, or removed entirely if all three coverage values are now above .75 + floating point epsilon.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy (hunk 12 above)
Comment (bakejeyner) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: Non-existent

Comment (blen-desta) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: It's not clear where you're doing that. Is that all the things you did in setup?

Comment (bakejeyner) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy]: I did do it in setup. Tried to clear it up with a comment.

**Comment by bakejeyner** [comment]
build now

File: platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<SamlProtocol.Binding> supportedBindings) {
+    AudienceBuilder audienceBuilder = new AudienceBuilder();
+    AudienceRestrictionBuilder audienceRestrictionBuilder = new AudienceRestrictionBuilder();
+
+    // According to the SAML spec, on an AuthnRequest, "[t]he <Issuer> element MUST be present and
+    // MUST contain the unique identifier of the requesting service provider".
+    Audience audience = buildAudience(audienceBuilder, authnRequest.getIssuer().getValue());
```
Comment (brjeter) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/main/java/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPlugin.java]: ✏️ I know this is a copy-paste of the original code, but can't we just new up an `AudienceBuilder` object inside the `buildAudience` method? From what I can tell, it is only used in that method.

**Comment by blen-desta** [comment]
Hero'd. Was able to log in.

**Pull request merged**